daemon/daemon/endpoint.go

```diff
@@ -202,42 +215,88 @@ func (d *Daemon) createBPFMAPs(epID uint16) error {
 	return ep.Regenerate(d)
 }
 
-// EndpointJoin sets up the endpoint working directory.
-func (d *Daemon) EndpointJoin(ep endpoint.Endpoint) error {
-	lxcDir := filepath.Join(".", strconv.Itoa(int(ep.ID)))
+type postEndpoint struct {
+	d *Daemon
+}
 
-	if err := os.MkdirAll(lxcDir, 0777); err != nil {
-		log.Warningf("Failed to create container temporary directory: %s", err)
-		return fmt.Errorf("failed to create temporary directory: %s", err)
+func NewPostEndpointHandler(d *Daemon) PostEndpointHandler {
+	return &postEndpoint{d: d}
+}
+
+func (h *postEndpoint) Handle(params PostEndpointParams) middleware.Responder {
+	ep, err := endpoint.NewEndpointFromModel(params.Endpoint)
+	if err != nil {
+		return apierror.Error(PostEndpointInvalidCode, err)
 	}
 
-	d.conf.OptsMU.RLock()
-	ep.SetDefaultOpts(d.conf.Opts)
-	d.conf.OptsMU.RUnlock()
+	h.d.endpointsMU.Lock()
+	defer h.d.endpointsMU.Unlock()
+	if _, ok := h.d.endpoints[ep.ID]; ok {
+		return NewPostEndpointExists()
+	}
 
-	d.InsertEndpoint(&ep)
+	ep.State = endpoint.StateCreated
+	h.d.insertEndpoint(ep)
+	// FIXME: Provide IPAM response
 
-	return nil
+	return NewPostEndpointCreated()
 }
 
-// EndpointLeave cleans the directory used by the endpoint epID and all relevant details
-// with the epID.
-func (d *Daemon) EndpointLeave(epID uint16) error {
-	d.endpointsMU.Lock()
-	defer d.endpointsMU.Unlock()
+type putEndpoint struct {
+	d *Daemon
+}
 
-	ep := d.lookupCiliumEndpoint(epID)
-	if ep == nil {
-		return fmt.Errorf("endpoint %d not found", epID)
+func NewPutEndpointIDHandler(d *Daemon) PutEndpointIDHandler {
+	return &putEndpoint{d: d}
+}
+
+func (h *putEndpoint) Handle(params PutEndpointIDParams) middleware.Responder {
+	ep, err := endpoint.NewEndpointFromModel(params.Endpoint)
+	if err != nil {
+		return apierror.Error(PutEndpointIDInvalidCode, err)
+	}
+
+	h.d.endpointsMU.Lock()
+	existingEp, ok := h.d.endpoints[ep.ID]
```

Comment:
True, this was unfinished, the new code is now locking throughout all changes. The existing code pre this PR operated on a copy which was racy as well. There was nothing to guarantee that no deletion request was processed between unlocking and locking again.